checks/cassandra/check.go

```diff
@@ -10,31 +10,37 @@ import (
 
 // Config is the Cassandra checker configuration settings container.
 type Config struct {
-	// Hosts is a list of Cassandra hosts. At least one is required.
+	// Hosts is a list of Cassandra hosts. Optional if Session is supplied.
 	Hosts []string
-	// Keyspace is the Cassandra keyspace to which you want to connect. Required.
+	// Keyspace is the Cassandra keyspace to which you want to connect. Optional if Session is supplied.
 	Keyspace string
+	// Session is a gocql session and can be used in place of Hosts and Keyspace. Recommended.
+	// Optional if Hosts & Keyspace are supplied.
+	Session *gocql.Session
 }
 
-// New creates new Cassandra health check that verifies the following:
-// - that a connection can be established through creating a session
-// - that queries can be executed by describing keyspaces
+// New creates new Cassandra health check that verifies that a connection exists and can be used to query the cluster.
 func New(config Config) func(ctx context.Context) error {
 	return func(ctx context.Context) error {
-		if len(config.Hosts) < 1 || len(config.Keyspace) < 1 {
-			return errors.New("keyspace name and hosts are required to initialize cassandra health check")
+		var session *gocql.Session
+		var err error
+		if config.Session == nil && (len(config.Hosts) < 1 || len(config.Keyspace) < 1) {
+			return errors.New("cassandra cluster config or keyspace name and hosts are required to initialize cassandra health check")
 		}
 
-		cluster := gocql.NewCluster(config.Hosts...)
-		cluster.Keyspace = config.Keyspace
+		if config.Session != nil {
+			session = config.Session
+		} else {
+			cluster := gocql.NewCluster(config.Hosts...)
+			cluster.Keyspace = config.Keyspace
+			session, err = cluster.CreateSession()
+			defer session.Close()
+		}
 
```

Comment:
This logic became quite complex and it's a bit unsafe to defer a method before checking the returned error, what do you think about extracting this part into a separate private function that safely returns what you need like this:

```golang
func initSession(c Config) (func(), *gocql.Session, error) {
	if c.Session != nil {
		return func() {
			// do not close provided session
		}, c.Session, nil
	}

	if len(c.Hosts) < 1 || len(c.Keyspace) < 1 {
		return nil, nil, errors.New("cassandra cluster config or keyspace name and hosts are required to initialize cassandra health check")
	}

	cluster := gocql.NewCluster(c.Hosts...)
	cluster.Keyspace = c.Keyspace
	session, err := cluster.CreateSession()
	if err != nil {
		return nil, nil, err
	}

	return session.Close, session, err
}
```

that way you can call it:
```golang
shutdown, session, err := initSession(config)
if err != nil {
	return fmt.Errorf("cassandra health check failed on connect: %w", err)
}
defer shutdown()
```

Reply:
Completely agreed. I refactored per your suggestions and tested it - all's working.